Gordon360/Services/HousingService.cs

```diff
@@ -531,4 +557,228 @@ public bool ChangeApplicationDateSubmitted(int applicationID)
         context.SaveChanges();
         return true;
     }
-}
+
+    /// <summary>
+    /// update roommate imformation
+    /// </summary>
+    /// <param name="username"> The username for the user who complete the aplication form </param>
+    /// <param name="application_id"> The ID of this application </param>
+    /// <param name="emailList"> A list of applicants' emails </param>
+    public async Task UpdateRoommateAsync(string username, string application_id, string[] emailList)
+    {
+        var account = accountService.GetAccountByUsername(username);
+        string gender = context.Student.FirstOrDefault(a => a.ID == account.GordonID).Gender;
+
+        var maxYear = context.Student
+            .Where(s => emailList.Contains(s.Email))
+            .Select(s => s.Class).Max();
+        await context.Year.AddAsync(new Year
+        {
+            ApplicationID = application_id,
+            Year1 = maxYear
+        });
+
+        foreach (string e in emailList)
+        {
+            if (e != "")
+            {
+                var existActiveApplicant = context.Applicant.FirstOrDefault(x => (x.Applicant1 == e) && (x.Active == 1));
+                if (existActiveApplicant != null)
+                {
+                    existActiveApplicant.Active = 0;
+                }
+                var student = context.Student.FirstOrDefault(x => x.Email == e);
+                if (student == null)
+                {
+                    throw new ResourceNotFoundException() { ExceptionMessage = "The applicant cannot be found." };
+                }
+                if (gender != student.Gender)
+                {
+                    throw new BadInputException() { ExceptionMessage = "The applicants are not of the same gender." };
+                }
+                gender = student.Gender;
+                var newApplicant = new Applicant
+                {
+                    ApplicationID = application_id,
+                    Applicant1 = e,
+                    Active = 1
+                }; ;
+                await context.Applicant.AddAsync(newApplicant);
+            }
+        }
+        context.SaveChanges();
+    }
+
+    /// <summary>
+    /// update the information of preferred halls
+    /// </summary>
+    /// <param name="username"> The username for the user who complete the aplication form </param>
+    /// <param name="application_id"> The ID of this application </param>
+    /// <param name="hallList"> A list of the preferred halls </param>
+    public void UpdatePreferredHall(string username, string application_id, string[] hallList)
+    {
+        var hallPreferences = hallList.Select((hall, index) => new PreferredHall { ApplicationID = application_id, Rank = index + 1, HallName = hall });
+        context.PreferredHall.AddRange(hallPreferences);
```

Comment:
Correct me if I'm wrong here, but it looks like you're adding without deleting former records, so this would be a create, not an update action.